### ISSUE TITLE
MS Teams Screensharing does not work

### ISSUE THREAD

**Issue by kekscode** [qutebrowser/qutebrowser#7320]
**Version info**:

MacOS
qutebrowser v2.5.2
Git commit: da10f19 on qt6-v2 (2022-07-15 16:26:10 +0200)
Backend: QtWebEngine 6.3.1, based on Chromium 94.0.4606.126 (from api)
Qt: 6.3.1

**Does the bug happen if you start with `--temp-basedir`?**:

Yes.

**Description**

During a call in MIcrosoft Teams (https://teams.microsoft.com/) Audio and Video works fine. But when trying to do a screenshare, the sharing tray gets displayed, but not the various screens/desktops to choose the actual source (the HTML widget just does not display). The same works in the official Chrome browser, though.

Generally screen sharing works fine (tested with: https://mozilla.github.io/webrtc-landing/gum_test.html)

It looks like an incompatibility with the MS Teams javascript code from Teams Web App. The web developer console gets various errors but i cannot make any sense of it and it is full of noise.

**How to reproduce**

- Go to https://teams.microsoft.com/ and log in (i don't know if there is a good way to test the behaviour without having to have an actual teams account?)
- Create a call and (try to) start screenshare in this call

**Comment by kekscode** [comment]
Oh maybe this is important: I think i had this issue already on the qt5 version. That's why i tried qt6 and the nightlies in the first place to check if this helps but i didn't.

**Comment by toofar** [comment]
Last I checked webengine doesn't support letting you select between windows/screens to share. It is hardcoded to share your primary screen on X11, no idea about wayland (https://bugreports.qt.io/browse/QTBUG-80054).

Does at least that work with teams? It's not clear to me from your report whether you can't share anything at all or whether you just can't choose.

FWIW there was a new release of the teams app for linux a few months ago and it works far better than previous ones.

**Comment by kekscode** [comment]
Hi @toofar thanks for the details. Two clarificationns:

- I am not on Linux, i use MacOS. Sorry if this was not clear based on the version info, i'll add this info to make it clearer
- I can't choose what to share (which screen/desktop). Sharing in general works technically e.g. with the mozilla gum_test.html

**Comment by The-Compiler** [comment]
Can you show a screenshot on how things look in Chrome vs. qutebrowser? It seems to work fine for me, but I don't see a HTML widget to select the screen, I get a native dialog (which is indeed missing in qutebrowser, it just shares the first screen).

Chrome:

![image](https://user-images.githubusercontent.com/625793/179472476-eca06f65-8ee1-4564-88e1-bf52bba3fc6a.png)

![image](https://user-images.githubusercontent.com/625793/179472514-833e0233-9eba-46cb-900e-e2a5c1ae4218.png)

While in qutebrowser, after the first click, I get a permission request, and then my first screen (laptop screen) gets shared successfully.

**Comment by kekscode** [comment]
### Chrome

This dialog opens in Chrome `Version 103.0.5060.114 (Official Build) (x86_64)` on MacOS:

![image](https://user-images.githubusercontent.com/803252/179477224-5b5e500d-2805-4099-958d-e4b9fe35cfa4.png)

It seems to fit pixel-perfect inside the page and has the same colorscheme, but it *may* be a native chrome widget though.

### Qutebrowser

I can see this so-called "share tray" but a click on "Desktop/Window" just does not trigger any action at all (also not a screen share).

![image](https://user-images.githubusercontent.com/803252/179477983-fc8b78b0-fb4d-41b8-a1aa-5068d37ff9ee.png)

I tested with Laptop-only and with my default setup (laptop closed and two displays connected).

**Comment by kekscode** [comment]
I get a lot of errors in the web console. One which seems to be at least loosely related is this:

```
2022-07-18T09:02:27.631Z [screenSharing][control] screenSharingControlService: unexpected, cannot find the call, teamsCallId [17]
```

and lots of traceback from minified js code with `0.2-angular-jquery.min-92954a0.js:114 TypeError: Cannot read properties of undefined (reading 'send')`

**Comment by The-Compiler** [comment]
I don't think those errors are relevant. I see the same when I share my screen in Chrome. That dialog is indeed a native dialog - you should see the exact same dialog on e.g. https://mozilla.github.io/webrtc-landing/gum_test.html (with Chromium). qutebrowser (or rather QtWebEngine) instead selects the first screen automatically.

I also just tried on macOS (though I can only test on macOS 10.14 right now), and I can indeed reproduce it there... Probably time to report this upstream! What macOS version are you on?

**Comment by kekscode** [comment]
OS: macOS Monterey 12.4
Model: MacBookPro15,1
CPU: 8-Core Intel Core i9
GPU: Radeon Pro 560X 4 GB
RAM: 16 GB

**Comment by The-Compiler** [comment]
Reported upstream here: [[QTBUG-105037] Screensharing does not work on MS Teams on macOS - Qt Bug Tracker](https://bugreports.qt.io/browse/QTBUG-105037)

Nothing more qutebrowser can do here, from what I can see.

**Comment by sga-13** [comment]
(sorry for posting on a closed issue)

I have a similar issue. When i try to do screenshare (i am on wayland), i get the display picker, and am able to select screen, but only a black screen comes. I am able to share screen on other webld beites, and also on teams in librewolf. I also checked the linked upstream bug, and that too is closed, so it seems that was sorted.

I have tried to run teams in temporary instance, so problem seemingly is not related to my config

I am using niri wm, and using xdg-desktop-portal-wlr for screen capture. I am using arch linux, and all packages are up to date as per arch repos. qt-webengine version is 6.9.2-2, and qutebrowser is 3.5.1, 

Any ideas what could be wrong?
